### PR TITLE
R client

### PR DESCRIPTION
**Pull request by lein-mckayla**
Hopefully this works! I don't understand the naming etiquette for commits so real sorry about that mess. I welcome any and all feedback. Thanks!

### CODE REVIEW

File: client/r/rhawc/R/rclientClass.R [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                      body = list(username = username, password = password), encode = "json")
+      token = (content(response))[[1]]
+      assign("token", token, thisEnv)
+    },
+    doSomethingAuthenticated = function(){
+      checkIsAuthenticated()
+      return(TRUE)
+    },
+    lit_import_hero = function(assessment_id, title, description, ids) {
+      url = paste0(root_url, "/lit/api/search/")
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: instead of paste0 everywhere, can we use glue instead? seems much more readable?

https://github.com/tidyverse/glue

```r
glue::glue("{root_url}/lit/api/search/")
```

File: client/r/rhawc/R/rclientClass.R [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      pc_json <- list(assessment = assessment_id,
+                      search_type = "i",
+                      source = 2,
+                      title = title,
+                      description = description,
+                      search_string = paste(unlist(ids), collapse=','))
+
+      response = POST(url,
+                      add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: let's make the adding the token header another private function that can be called from _get, _post, or anywhere else to reduce code duplication

File: hawc/apps/lit/templates/lit/overview.html [hunk 3]
```diff
@@ -59,6 +59,7 @@ <h3>Literature summary</h3>
             <th>Untagged</th>
             <td>{{overview.total_untagged}}</td>
           </tr>
+
```
Comment (shapiromatron) [inline hawc/apps/lit/templates/lit/overview.html]: can you revert the changes in this file? just delete the extra lines which were added.

File: hawc/apps/lit/api.py [hunk 4]
```diff
@@ -62,9 +62,11 @@ def reference_tags(self, request, pk):
         df = models.ReferenceTags.objects.as_dataframe(instance.id)
         return Response(df)
 
+
```
Comment (shapiromatron) [inline hawc/apps/lit/api.py]: can you revert the changes in this file? just delete the extra lines which were added.

File: client/r/rhawc/man/hello.Rd [hunk 5]
```diff
@@ -0,0 +1,12 @@
+\name{hello}
```
Comment (shapiromatron) [inline client/r/rhawc/man/hello.Rd]: do we need this file? if not, delete

File: client/r/rhawc/man/HawcClient.Rd [hunk 6]
```diff
@@ -0,0 +1,11 @@
+% Generated by roxygen2: do not edit by hand
```
Comment (shapiromatron) [inline client/r/rhawc/man/HawcClient.Rd]: do we need this file? if not, delete

File: client/r/rhawc/R/rclientClass.R [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  # the methods which will be exposed by the client
+  me <- list(
+    authenticate = function(username, password){
+      response = POST(paste0(root_url, "/user/api/token-auth/"),
+                      body = list(username = username, password = password), encode = "json")
+      token = (content(response))[[1]]
+      assign("token", token, thisEnv)
+    },
+    doSomethingAuthenticated = function(){
+      checkIsAuthenticated()
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: and remove this function too, we're not using it right?

File: client/r/rhawc/R/rclientClass.R [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+HawcClient <- function(baseUrl){
+
+  # local environment variables
+  thisEnv <- environment()
+
+  # persistent "state"
+  root_url <- baseUrl
+  #   token <- ""
+
+  checkIsAuthenticated <- function(){
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: remove `checkIsAuthenticated`? is it used?

File: client/r/rhawc/R/rclientClass.R [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    }
+    else if (get("token", thisEnv) == 'Unable to log in with provided credentials.') {
+      stop('Unable to log in with provided credentials.')
+    }
+  }
+  get_ = function(url) {
+    response = GET(url, add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
+                                    'Content-Type'="application/json"))
+    if (response$status_code >= 400 & response$status_code < 500) {
+      stop("An exception occured in the HAWC client module") }
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: can we print the response or something too before stopping? Perhaps as a `warning`? b/c often times there's an explanation to what caused the error in the response

File: client/r/rhawc/R/rclientClass.R [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+      response.list = content(response)
+      response.list = lapply(response.list, lapply, function(x)ifelse(is.null(x), NA, x))
+      response.df = rbind.fill(lapply(response.list, as.data.frame))
+      return(response.df) }
+  }
+  post_ = function(url, ...) {
+    response = POST(url, add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
+                                     'Content-Type'="application/json"), ...)
+    if (response$status_code >= 400 & response$status_code < 500) {
+      stop("An exception occured in the HAWC client module") }
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: same here, let's print the response as a warning so it's not lost

File: client/r/rhawc/R/rclientClass.R [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    response = GET(url, add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
+                                    'Content-Type'="application/json"))
+    if (response$status_code >= 400 & response$status_code < 500) {
+      stop("An exception occured in the HAWC client module") }
+
+    else if (response$status_code == 500) {
+      stop("An exception occured on the HAWC server") }
+
+    else {
+      response.list = content(response)
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: some responses should return a data frame, but not all.  in the python client, when it returns a Dict, here we should return the `response.list` and when it returns a datafarmae, we should return a dataframe. I'd make the get_ and post_ methods return a `response.list`, and then for some methods use that as.data.frame method to convert the list to a frame

File: client/r/rhawc/R/rclientClass.R [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  }
+  post_ = function(url, ...) {
+    response = POST(url, add_headers(Authorization=paste('Token', get("token", thisEnv),sep=' '),
+                                     'Content-Type'="application/json"), ...)
+    if (response$status_code >= 400 & response$status_code < 500) {
+      stop("An exception occured in the HAWC client module") }
+    else if (response$status_code == 500) {
+      stop("An exception occured on the HAWC server") }
+    else {
+      response.list = content(response)
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: same comment about response type here

File: client/r/rhawc/man/HawcClient.Rd (hunk 6 above)
Comment (shapiromatron) [reply client/r/rhawc/man/HawcClient.Rd]: if we do need some sort of docs, then let's just keep it short and copy the the package title name and put a link to our hawc documentation, https://hawc.readthedocs.org/

File: docs/source/api.rst [hunk 13]
```diff
@@ -41,3 +41,21 @@ There's also a HAWC client available:
         ids=hero_ids
```
Comment (shapiromatron) [inline docs/source/api.rst]: sorry, this comment doesn't belong on this file, but can you delete the `client/r/rhawc_0.1.0.tar.gz` file

File: client/r/rhawc/R/rclientClass.R [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+      response.list = lapply(response.list, lapply, function(x)ifelse(is.null(x), NA, x))
+      response.df = rbind.fill(lapply(response.list, as.data.frame))
+      return(response.df) }
+  }
+  # the methods which will be exposed by the client
+  me <- list(
+    authenticate = function(username, password){
+      response = POST(paste0(root_url, "/user/api/token-auth/"),
+                      body = list(username = username, password = password), encode = "json")
+      token = (content(response))[[1]]
```
Comment (shapiromatron) [inline client/r/rhawc/R/rclientClass.R]: can you please ensure this file is consistent with either using `=` or `<-`, since they're equivalent?

File: docs/source/api.rst [hunk 15]
```diff
@@ -41,3 +41,21 @@ There's also a HAWC client available:
...
     )
 
+There's also a HAWC client available in R:
+
+.. code-block:: R
+
+    install.packages("rhawc")
+    library(rhawc)
+
+    client = HawcClient("https://hawcproject.org")
```
Comment (shapiromatron) [inline docs/source/api.rst]: change these to `<-` or keep using `=`, depending on what you decide upon above

File: client/r/rhawc/.Rhistory [hunk 16]
```diff
@@ -0,0 +1,512 @@
+View(test)
```
[the remark this answers is not in the document]
Comment (lein-mckayla) [reply client/r/rhawc/.Rhistory]: i think i did this, but i only found one instance of .Rhistory

File: .gitignore [hunk 17]
```diff
@@ -42,3 +42,4 @@ hawc/static/bundles/
 
 compose/django/fonts/*
 !compose/django/fonts/.gitkeep
+.Rproj.user
```
[the remark this answers is not in the document]
Comment (lein-mckayla) [reply .gitignore]: Sorry I don't see a .gitignore file in this path, only .gitkeep, which I don't see any text in?

File: client/r/rhawc/DESCRIPTION [hunk 18]
```diff
@@ -0,0 +1,20 @@
+Package: rhawc
+Type: Package
+Title: A Tale of Translation, A HAWC Client 
+Version: 0.1.0
+Author: c(person("Mckayla", "Lein", email="mckayla.lein@gmail.com", role=c("cre", "cst", "dnc","spy")),
+    person("Daniel", "Rabstejnek", email="rabstejnek.daniel@epa.gov", role=c("aqt")))
+Maintainer: Mckayla Lein <mckayla.lein@gmail.com>
+Description: Doing cool things in the name of HAWC. This package is a knock off version of the Python client. 
+License: What license is it under?
```
[the remark this answers is not in the document]
Comment (lein-mckayla) [reply client/r/rhawc/DESCRIPTION]: resolved

File: client/r/rhawc/DESCRIPTION [hunk 19]
```diff
@@ -0,0 +1,20 @@
+Package: rhawc
+Type: Package
+Title: A Tale of Translation, A HAWC Client
```
[the remark this answers is not in the document]
Comment (lein-mckayla) [reply client/r/rhawc/DESCRIPTION]: resolved

File: .gitignore (hunk 17 above)
[the remark this answers is not in the document]
Comment (shapiromatron) [reply .gitignore]: it should be at the root, so `~/dev/hawc/.gitignore`

File: client/r/rhawc/.Rhistory (hunk 16 above)
[the remark this answers is not in the document]
Comment (shapiromatron) [reply client/r/rhawc/.Rhistory]: one here too `hawc/client/.Rhistory`; I'm looking at the files on the "files changed" tab above

File: hawc/apps/lit/templates/lit/overview.html (hunk 3 above)
Comment (lein-mckayla) [reply hawc/apps/lit/templates/lit/overview.html]: Sorry but I don't understand what you mean here? I'm not sure which extra lines were added, unless you mean just delete white space?

File: hawc/apps/lit/api.py (hunk 4 above)
Comment (lein-mckayla) [reply hawc/apps/lit/api.py]: This goes with above. I know Danny and I opened these files and did something a yesterday when I was having trouble accessing the API urls, and I was unable to automatically merge the master branch with mine. I don't really understand it though, should it just match what's in the master branch for this file?

File: hawc/apps/lit/templates/lit/overview.html (hunk 3 above)
Comment (shapiromatron) [reply hawc/apps/lit/templates/lit/overview.html]: the green lines in the diffs were changes that you made to these files; it looks like you added some whitespace in this file and another file. you can delete those lines so that when we look at the "file changes" tab these files no longer appear

File: client/r/rhawc/man/HawcClient.Rd (hunk 6 above)
Comment (lein-mckayla) [reply client/r/rhawc/man/HawcClient.Rd]: I do not know if this is 100% necessary, but it is a file that is automatically generated during package development (and updated every time i update the package) and as far as I know every package I've seen has one for all the functions. I believe it is what take the description information at the top of an .R file (i.e. lines that lead with #') and turns them into the help documentation that is available when you type ?function or help(function). Maybe you already knew all that, in which case I will look into how to get rid of it.

File: hawc/apps/lit/api.py (hunk 4 above)
Comment (shapiromatron) [reply hawc/apps/lit/api.py]: just delete the blank lines that were added here, so that when you push the changes up the file disappears from this "diff" view

Comment (shapiromatron) [reply hawc/apps/lit/api.py]: Also, I just realized I don't know how to build an r package - can you add some instructions for how to build the package in a new section in `docs/development.rst`?

File: client/r/rhawc/DESCRIPTION [hunk 20]
```diff
@@ -0,0 +1,20 @@
+Package: rhawc
+Type: Package
+Title: HAWC (Health assessment workspace collaborative) client 
+Version: 0.1.0
+Author: c(person("Mckayla", "Lein", email="mckayla.lein@gmail.com", role=c("cre", "cst", "dnc","spy")),
+    person("Daniel", "Rabstejnek", email="rabstejnek.daniel@epa.gov", role=c("aqt")))
+Maintainer: Mckayla Lein <mckayla.lein@gmail.com>
+Description: Doing cool things in the name of HAWC. This package is a knock off version of the Python client.
```
Comment (shapiromatron) [inline client/r/rhawc/DESCRIPTION]: change the description to "An R-client to send and receive data from HAWC (Health assessment workspace collaborative).`

File: client/r/rhawc/man/HawcClient.Rd (hunk 6 above)
Comment (shapiromatron) [reply client/r/rhawc/man/HawcClient.Rd]: nope, I wasn't sure. If we need it then we'll keep it. Can you update the description per my comment above?
